src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -2002,11 +2002,14 @@ public void observe(@NonNull SCMHead head, @NonNull SCMRevision revision) throws
                         // either because the head was removed and then recreated, or because the head
                         // was taken over by a different source, thus the previous revision is null
                         doAutomaticBuilds(head, revision, rawName, project, revisionActions, null, null);
-                    } else if (revision.isDeterministic()) {
+                    } else {
+                        // get the previous revision
                         SCMRevision scmLastBuiltRevision = _factory.getRevision(project);
-                        SCMRevision scmLastSeenRevision = getLastSeenRevision(project, scmLastBuiltRevision);
-                        if (!revision.equals(scmLastBuiltRevision)) {
+
+                        if (isChangesDetected(revision, project, scmLastBuiltRevision)) {
```

Comment:
Instead of running two almost identical blocks, the checks are handled by `isChangesDetected` and then the actual same code path is executed.